[JSoC'20] Multithreaded Connected Components

abhinavmehndiratta opened the pull request:
The speedup achieved by the parallel version is about `nthreads/log(V)`. I could only test on 6 cores, but it should perform better with more threads.

```julia
julia> using LightGraphs, BenchmarkTools

julia> using LightGraphs.Connectivity

julia> using MatrixDepot

julia> md = mdopen("DIMACS10/kron_g500-logn16")

julia> g = Graph(md.A)
{65536, 2456398} undirected simple Int64 graph

julia> x = Connectivity.connected_components(g)

julia> y = Connectivity.connected_components(g, Connectivity.PointerJumping())

julia> x == y
true

julia> @btime Connectivity.connected_components(g)
  25.210 ms (10280 allocations: 4.04 MiB)
```
Parallel version -

1 thread -
```julia
julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  143.793 ms (10702 allocations: 70.08 MiB)
```
2 threads -
```julia
julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  73.707 ms (10928 allocations: 71.10 MiB)
```
4 threads -
```julia
julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  54.085 ms (11464 allocations: 73.15 MiB)
```
6 threads -
```julia
julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  50.008 ms (12060 allocations: 59.70 MiB)
```

sbromberger commented:
At 6 threads it's still twice as slow as the single-threaded default?

sbromberger commented on src/Connectivity/Connectivity.jl:
```diff
@@ -2,13 +2,16 @@ module Connectivity
 import LightGraphs.Traversals: initfn!, newvisitfn!, postlevelfn!, visitfn!, previsitfn!
+using Base.Threads
+import LightGraphs.optimal_contiguous_partition
+import LightGraphs.SimpleGraphs: SimpleGraph, SimpleDiGraph, SimpleEdge
```
Library code must only use SimpleGraphsCore.

abhinavmehndiratta commented:
Updated benchmark (6 threads) -

```julia
julia> g = loadsnap(:facebook_combined)
{4039, 88234} undirected simple Int64 graph

julia> @btime Connectivity.connected_components(g)
  750.548 μs (27 allocations: 160.25 KiB)

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  659.018 μs (608 allocations: 345.55 KiB)

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping(contract_edges = true))
  1.563 ms (1144 allocations: 6.39 MiB)
```

```julia
julia> g = loadsnap(:soc_slashdot0902_u)
{82168, 582533} undirected simple Int64 graph

julia> @btime Connectivity.connected_components(g)
  11.389 ms (32 allocations: 3.88 MiB)

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  18.848 ms (647 allocations: 6.45 MiB)

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping(contract_edges = true))
  15.358 ms (1171 allocations: 36.50 MiB)
```
Updated benchmark on the same graph I used above

```julia
julia> md = mdopen("DIMACS10/kron_g500-logn16")

julia> g = Graph(md.A)
{65536, 2456398} undirected simple Int64 graph

julia> @btime Connectivity.connected_components(g)
  24.617 ms (10280 allocations: 4.04 MiB)         

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  30.854 ms (10861 allocations: 6.10 MiB)                                                                                 

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping(contract_edges = true))
  38.033 ms (11429 allocations: 108.15 MiB)      
```

Note :

Without edge reduction - the speedup in this case is almost fixed, around p/log(V). So for a million vertices, you'll need about 19 cores to see a speedup. 
This algorithm doesn't use any extra memory for storing edges since we scan the same edges in every iteration.

With edge reduction - the speedup in this case depends upon the distribution of the graph. At each step, the number of vertices is halved (by hooking to parent) but it is NOT necessary that the number of edges is also halved. It will generally give more speedup, however the problem here is load imbalance. It works well with very dense and very sparse graphs. But for cases where the graph has a medium density with a giant connected component, the algorithm will perform poorly. This is because for a graph with a single large component, many roots will hook to the same component, resulting in load imbalance and the extra step of edge contraction is rendered useless.
This algorithm uses extra memory for storing edges since the edges are reduced after each iteration.

This is the only well performing algorithm I could find and I've tried to optimize it as much as I could.

abhinavmehndiratta commented:
More benchmarks on small graphs -
```julia
julia> Base.Threads.nthreads()
6

julia> md = mdopen("Cylshell/s1rmt3m1")

julia> g = Graph(md.A)
{5489, 111570} undirected simple Int64 graph

julia> @btime Connectivity.connected_components(g)
  881.450 μs (28 allocations: 258.44 KiB)

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  823.969 μs (609 allocations: 489.23 KiB)

julia> md = mdopen("VDOL/hangGlider_3")

julia> g = Graph(md.A)
{10260, 49201} undirected simple Int64 graph

julia> @btime Connectivity.connected_components(g)
  527.508 μs (29 allocations: 498.25 KiB)

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  514.528 μs (434 allocations: 859.97 KiB)
```

sbromberger commented:
I'll try this on a 64-core machine next week.

abhinavmehndiratta commented:
> I'll try this on a 64-core machine next week.

Awesome. Thanks.

sbromberger commented:
Here are some benchmarks on a 64-core system:
```
julia> g = loadsnap(:soc_slashdot0902_u)
{82168, 582533} undirected simple Int64 graph
julia> @benchmark Connectivity.connected_components(g)
BenchmarkTools.Trial: 
  memory estimate:  3.88 MiB
  allocs estimate:  32
  --------------
  minimum time:     19.140 ms (0.00% GC)
  median time:      20.905 ms (0.00% GC)
  mean time:        22.656 ms (0.57% GC)
  maximum time:     37.704 ms (4.52% GC)
  --------------
  samples:          220
  evals/sample:     1

julia> @benchmark Connectivity.connected_components(g, Connectivity.PointerJumping())
BenchmarkTools.Trial: 
  memory estimate:  6.97 MiB
  allocs estimate:  4263
  --------------
  minimum time:     74.994 ms (0.00% GC)
  median time:      83.332 ms (0.00% GC)
  mean time:        84.269 ms (0.41% GC)
  maximum time:     99.749 ms (4.35% GC)
  --------------
  samples:          60
  evals/sample:     1

julia> @benchmark Connectivity.connected_components(g, Connectivity.PointerJumping(contract_edges=true))
BenchmarkTools.Trial: 
  memory estimate:  71.46 MiB
  allocs estimate:  8460
  --------------
  minimum time:     82.289 ms (0.00% GC)
  median time:      144.886 ms (0.00% GC)
  mean time:        142.872 ms (0.80% GC)
  maximum time:     166.385 ms (9.71% GC)
  --------------
  samples:          35
  evals/sample:     1
```

sbromberger commented:
Better performance with 32 threads:
```
julia> @benchmark Connectivity.connected_components(g)
BenchmarkTools.Trial: 
  memory estimate:  3.88 MiB
  allocs estimate:  32
  --------------
  minimum time:     18.452 ms (0.00% GC)
  median time:      20.215 ms (0.00% GC)
  mean time:        20.423 ms (0.62% GC)
  maximum time:     36.294 ms (0.00% GC)
  --------------
  samples:          244
  evals/sample:     1
julia> @benchmark Connectivity.connected_components(g, Connectivity.PointerJumping())
BenchmarkTools.Trial: 
  memory estimate:  6.68 MiB
  allocs estimate:  2172
  --------------
  minimum time:     20.860 ms (0.00% GC)
  median time:      30.171 ms (0.00% GC)
  mean time:        31.562 ms (0.82% GC)
  maximum time:     60.652 ms (2.68% GC)
  --------------
  samples:          158
  evals/sample:     1


julia> @benchmark Connectivity.connected_components(g, Connectivity.PointerJumping(contract_edges=true))
BenchmarkTools.Trial: 
  memory estimate:  70.93 MiB
  allocs estimate:  4314
  --------------
  minimum time:     28.266 ms (0.00% GC)
  median time:      32.737 ms (0.00% GC)
  mean time:        36.705 ms (0.00% GC)
  maximum time:     59.144 ms (0.00% GC)
  --------------
  samples:          116
  evals/sample:     1

julia> versioninfo()
Julia Version 1.4.1
Commit 381693d3df* (2020-04-14 17:20 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-4610 v2 @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, ivybridge)
Environment:
  JULIA_NUM_THREADS = 32
```

sbromberger commented:
17 (?!) threads:
```
julia> @benchmark Connectivity.connected_components(g)
BenchmarkTools.Trial: 
  memory estimate:  3.88 MiB
  allocs estimate:  32
  --------------
  minimum time:     20.489 ms (0.00% GC)
  median time:      29.604 ms (0.00% GC)
  mean time:        29.874 ms (0.61% GC)
  maximum time:     80.059 ms (0.00% GC)
  --------------
  samples:          167
  evals/sample:     1
julia> @benchmark Connectivity.connected_components(g, Connectivity.PointerJumping())
BenchmarkTools.Trial: 
  memory estimate:  6.55 MiB
  allocs estimate:  1207
  --------------
  minimum time:     19.454 ms (0.00% GC)
  median time:      28.197 ms (0.00% GC)
  mean time:        29.522 ms (0.88% GC)
  maximum time:     45.158 ms (0.00% GC)
  --------------
  samples:          169
  evals/sample:     1
julia> @benchmark Connectivity.connected_components(g, Connectivity.PointerJumping(contract_edges=true))
BenchmarkTools.Trial: 
  memory estimate:  40.68 MiB
  allocs estimate:  2350
  --------------
  minimum time:     14.540 ms (0.00% GC)
  median time:      22.518 ms (0.00% GC)
  mean time:        24.621 ms (7.40% GC)
  maximum time:     59.505 ms (42.38% GC)
  --------------
  samples:          199
  evals/sample:     1
```

sbromberger commented:
8 threads:
```

julia> @benchmark Connectivity.connected_components(g)
BenchmarkTools.Trial: 
  memory estimate:  3.88 MiB
  allocs estimate:  32
  --------------
  minimum time:     18.150 ms (0.00% GC)
  median time:      19.239 ms (0.00% GC)
  mean time:        19.567 ms (0.48% GC)
  maximum time:     34.535 ms (0.00% GC)
  --------------
  samples:          255
  evals/sample:     1
julia> @benchmark Connectivity.connected_components(g, Connectivity.PointerJumping())
BenchmarkTools.Trial: 
  memory estimate:  6.46 MiB
  allocs estimate:  615
  --------------
  minimum time:     16.714 ms (0.00% GC)
  median time:      19.547 ms (0.00% GC)
  mean time:        20.570 ms (1.42% GC)
  maximum time:     46.251 ms (35.42% GC)
  --------------
  samples:          243
  evals/sample:     1
julia> @benchmark Connectivity.connected_components(g, Connectivity.PointerJumping(contract_edges=true))
BenchmarkTools.Trial: 
  memory estimate:  46.53 MiB
  allocs estimate:  1196
  --------------
  minimum time:     22.116 ms (0.00% GC)
  median time:      30.163 ms (0.00% GC)
  mean time:        31.551 ms (6.34% GC)
  maximum time:     53.749 ms (45.60% GC)
  --------------
  samples:          159
  evals/sample:     1
```

abhinavmehndiratta commented:
Turns out I misunderstood the research paper, I've corrected the algorithm and faster than serial in every case and I've completely removed edge contraction, because it was ineffective in all my tests. 

Updated benchmarks -
```julia
julia> Base.Threads.nthreads()
6

julia> g = loadsnap(:facebook_combined)
{4039, 88234} undirected simple Int64 graph

julia> x = Connectivity.connected_components(g, Connectivity.PointerJumping());

julia> y = Connectivity.connected_components(g);

julia> x == y
true

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  425.859 μs (253 allocations: 277.79 KiB)

julia> @btime Connectivity.connected_components(g)
  749.728 μs (27 allocations: 160.25 KiB)

julia> md = mdopen("DIMACS10/kron_g500-logn16")

julia> g = Graph(md.A)
{65536, 2456398} undirected simple Int64 graph

julia> x = Connectivity.connected_components(g, Connectivity.PointerJumping());

julia> y = Connectivity.connected_components(g);

julia> x == y
true

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  15.144 ms (10506 allocations: 5.57 MiB)
                                                                                             
julia> @btime Connectivity.connected_components(g)
  24.897 ms (10280 allocations: 4.04 MiB)

julia> g = loadsnap(:ego_twitter_u)
{81306, 1342310} undirected simple Int64 graph

julia> x = Connectivity.connected_components(g, Connectivity.PointerJumping());

julia> y = Connectivity.connected_components(g);

julia> x == y
true

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  12.809 ms (350 allocations: 5.76 MiB)

julia> @btime Connectivity.connected_components(g)
  18.849 ms (32 allocations: 3.86 MiB)

julia> g = loadsnap(:soc_slashdot0902_u)
{82168, 582533} undirected simple Int64 graph

julia> x = Connectivity.connected_components(g, Connectivity.PointerJumping());

julia> y = Connectivity.connected_components(g);

julia> x == y
true

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  10.185 ms (262 allocations: 5.79 MiB)

julia> @btime Connectivity.connected_components(g)
  13.383 ms (32 allocations: 3.88 MiB)
```

sbromberger commented:
awesome work!

abhinavmehndiratta commented:
Benchmark on 5 million edges -
```julia
julia> using LightGraphs.Connectivity

julia> using MatrixDepot

julia> Base.Threads.nthreads()
6

julia> md = mdopen("DIMACS10/kron_g500-logn17")
(IS DIMACS10/kron_g500-logn17(#2490)  131072x131072(10228360/5114375) 2010 [A] 'Undirected Multigraph' [DIMACS10 set: kronecker/kron_g500-logn17]()

julia> g = Graph(md.A)
{131072, 5114375} undirected simple Int64 graph

julia> x = Connectivity.connected_components(g, Connectivity.PointerJumping());                                                               

julia> y = Connectivity.connected_components(g);                                                                                                     

julia> x == y
true

julia> @btime Connectivity.connected_components(g, Connectivity.PointerJumping())
  32.036 ms (23556 allocations: 12.07 MiB)

julia> @btime Connectivity.connected_components(g)
  54.397 ms (23242 allocations: 9.04 MiB)                                                                  
```

sbromberger commented:
@abhinavmehndiratta is this ready for review and merge?

sbromberger commented on src/Connectivity/pointer_jumping.jl:
```diff
@@ -0,0 +1,48 @@
+based on parallel opportunistic pointer jumping
+
+### References
+Connected-Components algorithms for Mesh-Connected Parallel Computers
```
bullet the references, and give a URL if you have one.

sbromberger commented on src/Connectivity/pointer_jumping.jl:
```diff
@@ -0,0 +1,48 @@
+struct PointerJumping <: WeakConnectivityAlgorithm end
+
+function connected_components(g::SimpleGraph{T}, alg::PointerJumping) where T <: Integer
+    nthrds = nthreads()
```
why do you assign a variable when you use `nthreads()` exactly once?

sbromberger commented on src/Connectivity/pointer_jumping.jl:
```diff
@@ -0,0 +1,48 @@
+    while oldP != P
+        # opportunistic pointer jumping
+        @threads for u_set in partitions
+            tid = threadid()
```
where are you using `tid`?

sbromberger commented on test/Connectivity/connectivity.jl:
```diff
@@ -27,6 +27,17 @@
+            @test cc[5] == [12]
+        end
+    end
+
```
Check also against the output of a non-threaded `connected_components`

abhinavmehndiratta commented:
> @abhinavmehndiratta is this ready for review and merge?

Yes. I've made the changes.

sbromberger commented on src/Connectivity/pointer_jumping.jl:
```diff
@@ -0,0 +1,47 @@
+"""
+struct PointerJumping <: WeakConnectivityAlgorithm end
+
+function connected_components(g::SimpleGraph{T}, alg::PointerJumping) where T <: Integer
```
Why is this not `AbstractGraph{T}`?

If it works with `<:AbstractGraph` then you can remove the `SimpleGraph` import in `Connectivity.jl`.

sbromberger commented on src/Connectivity/Connectivity.jl:
```diff
@@ -2,13 +2,15 @@ module Connectivity
 using SimpleTraits
 import LightGraphs.Traversals: initfn!, newvisitfn!, postlevelfn!, visitfn!, previsitfn!
+using Base.Threads
+import LightGraphs.greedy_contiguous_partition
```
this should be `using` since we're not adding methods.

sbromberger commented on src/Connectivity/Connectivity.jl:
```diff
@@ -309,5 +311,6 @@ include("kosaraju.jl")
 include("neighborhood_dists.jl")
 include("tarjan.jl")
 include("unionmerge.jl")
+include("pointer_jumping.jl")
```
rename to `threaded-pointer_jumping` please - keeps it consistent.

sbromberger commented on src/Connectivity/pointer_jumping.jl:
```diff
@@ -0,0 +1,47 @@
+"""
+    struct PointerJumping <: WeakConnectivityAlgorithm
```
rename to ThreadedPointerJumping please.

sbromberger commented:
Looks great. Thanks!
